java/core/src/java/org/apache/orc/impl/ColumnStatisticsImpl.java

```diff
@@ -993,8 +993,7 @@ public void updateBinary(byte[] bytes, int offset, int length,
 
     @Override
     public void merge(ColumnStatisticsImpl other) {
-      if (other instanceof BinaryColumnStatistics) {
-        BinaryStatisticsImpl bin = (BinaryStatisticsImpl) other;
+      if (other instanceof BinaryStatisticsImpl bin) {
```

Comment:
Actually, this PR is still not robust. This could be worse than before.

Instead of CCE, the corner case slips though `else` statement. And, If `isStatsExists() && sum != 0` is false, it propagates to `super.merge(other)`.

Reply:
It seems that other types of merge methods have similar problems. When itself is empty, it will merge statistics of incompatible types. I want to open a new JIRA to fix this problem. What do you think?

Here's a test that passes the test.
```java
    TypeDescription stringSchema = TypeDescription.createString();
    ColumnStatisticsImpl stringStats = ColumnStatisticsImpl.create(stringSchema);

    TypeDescription doubleSchema = TypeDescription.createDouble();
    ColumnStatisticsImpl doubleStats = ColumnStatisticsImpl.create(doubleSchema);

    stringStats.increment(3);
    stringStats.updateString(new Text("bob"));
    stringStats.updateString(new Text("david"));
    stringStats.updateString(new Text("charles"));

    doubleStats.merge(stringStats);
    assertEquals(3, ((DoubleColumnStatistics) doubleStats).getNumberOfValues());
```

Reply:
Thank you for looking at that. How many JIRAs do you want to handle in total? If everything is due to the same pattern, I'd put it together in ORC-1683 (this).